### PR TITLE
Refactor for data.table

### PR DESCRIPTION
**Pull request by tanho63**

### CODE REVIEW

File: R/from_url.R [hunk 1]
```diff
@@ -20,6 +20,7 @@ rds_from_url <- function(url){
     return(data.frame())
```
Comment (mrcaseb) [inline R/from_url.R:20]: Should this also return a data.table?

File: R/from_url.R [hunk 2]
```diff
@@ -20,6 +20,7 @@ rds_from_url <- function(url){
...
   }
 
+  data.table::setDT(load)
```
Comment (mrcaseb) [inline R/from_url.R:23]: So we are basically returning a data.table reference now. I am not used to that concept. Is this safe to do because we also export that function for the users?

File: R/load_nflverse.R [hunk 3]
```diff
@@ -35,15 +35,16 @@ load_pbp <- function(seasons = most_recent_season(), file_type = getOption("nflr
   p <- NULL
   if (is_installed("progressr")) p <- progressr::progressor(along = seasons)
 
-  out <- purrr::map_dfr(urls, progressively(loader, p))
-  class(out) <- c("tbl_df","tbl","data.frame")
+  out <- lapply(urls, progressively(loader, p))
```
Comment (mrcaseb) [inline R/load_nflverse.R:38]: just fyi I have tested parallel::mclapply again and it reduced all pbp runtime from 36 seconds to 26 seconds now that we have an efficient way to bind the lists. Might be worthwile to check that in another PR

File: R/load_nflverse.R [hunk 4]
```diff
@@ -70,13 +71,19 @@ load_player_stats <- function(seasons = most_recent_season(),
...
+
   loader <- choose_loader(file_type)
 
-  url <- paste0("https://github.com/nflverse/nflfastR-data/raw/master/data/player_stats.",file_type)
+  url <- paste0("https://github.com/nflverse/nflfastR-data/raw/master/data/",base_name,file_type)
 
   out <- loader(url)
-  class(out) <- c("tbl_df","tbl","data.frame")
-  if(!is.null(seasons)) out <- dplyr::filter(out, out$season %in% seasons)
+  if(!is.null(seasons)) out <- out[out$season %in% seasons,]
```
Comment (mrcaseb) [inline R/load_nflverse.R:85]: I am quite sure this is not recognized as data.table syntax because it never gets imported (except DESCRIPTION). I would suggest adding this to `R/nflreadr-package.R`

```r
# Make sure data.table knows we know we're using it
.datatable.aware = TRUE
```
See https://rdatatable.gitlab.io/data.table/articles/datatable-importing.html#data-table-in-imports-but-nothing-imported

Comment (mrcaseb) [reply R/load_nflverse.R:85]: I also realized that the `if` statement is nonsense since seasons is never `NULL` That line is executed every time

File: R/load_nflverse.R [hunk 5]
```diff
@@ -209,8 +217,8 @@ load_nextgen_stats <- function(seasons = TRUE,
                 file_type)
 
   out <- loader(url)
-  class(out) <- c("tbl_df","tbl","data.frame")
-  if(!is.null(seasons)) out <- dplyr::filter(out, out$season %in% seasons)
+  if(!is.null(seasons)) out <- out[out$season %in% seasons,]
```
Comment (mrcaseb) [inline R/load_nflverse.R:220]: See `load_player_stats` comments

File: R/load_nflverse.R [hunk 6]
```diff
@@ -294,8 +303,8 @@ load_injuries <- function(seasons = most_recent_season(),
                 ".",
                 file_type)
   out <- loader(url)
-  class(out) <- c("tbl_df","tbl","data.frame")
-  if(!is.null(seasons)) out <- dplyr::filter(out, out$season %in% seasons)
+  if(!is.null(seasons)) out <- out[out$season %in% seasons,]
```
Comment (mrcaseb) [inline R/load_nflverse.R:306]: See `load_player_stats` comments

File: R/load_nflverse.R [hunk 7]
```diff
@@ -335,9 +344,8 @@ load_espn_qbr <- function(league = c("nfl", "college"),
                 ".rds")
 
   out <- rds_from_url(url)
-  class(out) <- c("tbl_df","tbl","data.frame")
-
-  if(!is.null(seasons)) out <- dplyr::filter(out, out$season %in% seasons)
+  if(!is.null(seasons)) out <- out[out$season %in% seasons,]
```
Comment (mrcaseb) [inline R/load_nflverse.R:347]: See `load_player_stats` comments

File: R/load_nflverse.R [hunk 8]
```diff
@@ -371,8 +379,8 @@ load_pfr_passing <- function(seasons = TRUE){
   url <- "https://raw.githubusercontent.com/nflverse/pfr_scrapR/master/data/pfr_advanced_passing.rds"
 
   out <- rds_from_url(url)
-  class(out) <- c("tbl_df","tbl","data.frame")
-  if(!is.null(seasons)) out <- dplyr::filter(out, out$season %in% seasons)
+  if(!is.null(seasons)) out <- out[out$season %in% seasons,]
```
Comment (mrcaseb) [inline R/load_nflverse.R:382]: See `load_player_stats` comments

File: R/load_nflverse.R [hunk 9]
```diff
@@ -433,10 +443,9 @@ load_draft_picks <- function(seasons = TRUE){
 
   url <- "https://raw.githubusercontent.com/nflverse/nfldata/master/data/draft_picks.rds"
   out <- rds_from_url(url)
-  class(out) <- c("tbl_df","tbl","data.frame")
-
   if(!isTRUE(seasons)) stopifnot(is.numeric(seasons))
-  if(!isTRUE(seasons)) out <- dplyr::filter(out, out$season %in% seasons)
+  if(!isTRUE(seasons)) out <- out[out$season %in% seasons,]
```
Comment (mrcaseb) [inline R/load_nflverse.R:447]: See `load_player_stats` comments

File: R/load_nflverse.R (hunk 4 above)
Comment (mrcaseb) [reply R/load_nflverse.R:85]: I think the `out$` syntax won't be recognized as data.table syntax at all. See the following example

It looks like we have to get rid of `out$` and define `season <- NULL` globally somewhere in the package to avoid R CMD NOTES

``` r
library(data.table)

flights <- fread("https://raw.githubusercontent.com/Rdatatable/data.table/master/vignettes/flights14.csv")

b <- bench::mark(
  dt_filt = flights[origin == "JFK" & month == 6L],
  filt = flights[flights$origin == "JFK" & flights$month == 6L,],
  min_iterations = 500
)

ggplot2::autoplot(b) 
#> Loading required namespace: tidyr
```

File: R/from_url.R (hunk 1 above)
Comment (tanho63) [reply R/from_url.R:20]: Yes :)

File: R/from_url.R (hunk 2 above)
Comment (tanho63) [reply R/from_url.R:23]: We're returning a data.frame with some extra data.table attributes - it's "safe" not because we're exporting it but because R/tidyverse/data.table etc will convert it to whatever the user needs in order to do their operations

File: R/load_nflverse.R (hunk 3 above)
Comment (tanho63) [reply R/load_nflverse.R:38]: 👍

File: R/load_nflverse.R (hunk 4 above)
Comment (tanho63) [reply R/load_nflverse.R:85]: NULL bit definitely a holdover from when I initially wrote it with NULL = no filter

Comment (tanho63) [reply R/load_nflverse.R:85]: Unsure I can replicate your benchmark
```
R> pkgload::load_all()
i Loading nflreadr
R> library(data.table)
R> ps <- load_player_stats(TRUE)
R> setDT(ps)
R> bench::mark(
+++   ps[ps$season %in% 2019:2020,],
+++   ps[season %in% 2019:2020,],
+++   ps[season %in% 2019:2020],
+++   min_iterations = 500
+++ )
# A tibble: 3 x 13
  expression                         min  median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time
  <bch:expr>                     <bch:t> <bch:t>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm>
1 ps[ps$season %in% 2019:2020, ]  2.99ms  3.25ms      302.    4.68MB     35.8   447    53      1.48s
2 ps[season %in% 2019:2020, ]     3.46ms  3.92ms      249.    4.01MB     18.7   465    35      1.87s
3 ps[season %in% 2019:2020]       3.44ms  3.74ms      257.    3.15MB     15.2   472    28      1.84s
# ... with 4 more variables: result <list>, memory <list>, time <list>, gc <list>
```

Comment (tanho63) [reply R/load_nflverse.R]: adding the datatable aware

Comment (mrcaseb) [reply R/load_nflverse.R]: Oh. The benchmark was taken from an data.table example and I didn't test it without the `&` operator. I agree that the difference is much smaller without an `&`. That's wild but who cares. Your code should be fine

**Review by mrcaseb — APPROVED** [review]
